c7n/resources/redshift.py

```diff
@@ -603,6 +604,103 @@ def process(self, clusters):
         return clusters
 
 
+@Redshift.action_registry.register('set-attributes')
+class RedshiftSetAttributes(BaseAction):
+    """
+    Action to modify Redshift clusters
+    :example:
+    .. code-block:: yaml
+            policies:
+                - name: redshift-modify-cluster
+                  resource: redshift
+                  filters:
+                    - type: value
+                      key: AllowVersionUpgrade
+                      value: false
+                  actions:
+                    - type: set-attributes
+                      attributes:
+                        AllowVersionUpgrade: true
+    """
+
+    schema = type_schema('set-attributes',
+                        attributes={"type": "object"},
+                        required=('attributes',))
+
+    permissions = ('redshift:ModifyCluster',)
+
+    shape = 'ModifyClusterMessage'
+
+    def validate(self):
+        attrs = dict(self.data.get('attributes'))
+        if attrs.get('ClusterIdentifier'):
+            raise PolicyValidationError('ClusterIdentifier field cannot be updated')
+        attrs["ClusterIdentifier"] = ""
+        return shape_validate(attrs, self.shape, 'redshift')
+
+    def process(self, clusters):
+        client = local_session(self.manager.session_factory).client(
+            self.manager.get_model().service)
+        for cluster in clusters:
+            self.process_cluster(client, cluster)
+
+    def process_cluster(self, client, cluster):
+        try:
+            config = dict(self.data.get('attributes'))
+            base_config = {
```

Comment:
some of this logic would be simplified quite a bit by using jmespath to extract the nested values.

Reply:
Ok, should look a bit better now

Reply:
let's take a step back.. why do you this base config have this at all?

Reply:
The base_config gets the data given by describe_clusters and formats it so that it is equivalent to the format of the modify_cluster boto call. This way I can check to see if any attributes passed in through the policy are different from the current attributes of the cluster (lines 690-692). If there are differences, I can call modify cluster with these updated attributes (line 696). I also want to prevent making duplicate modify_cluster calls, which is why I prefer getting data from PendingModifiedValues (don't need to update a cluster if that update is already pending).

Reply:
but the keys are the same, so just iterating through the user set options, and checking them against pending modified, and cluster config would reduce the big static dictionary def, and reduce the code size on this considerably as well being more future proof.

Reply:
ie, this would save about 50 lines, be simpler and more flexible

```python
modify = {}
for k, v in user_set.items():
    if v != cluster_pending.get(k, cluster.get(k)):
         modify[k] = v
if not modify:
    continue # return
```

also if there are pending mods, does redshift allow additional modifies?

Reply:
Your solution is a lot cleaner, but I don't think it fully works because there are still values in HsmStatus, ElasticIpStatus, ClusterSecurityGroups, and VpcSecurityGroups that are not in cluster_pending (ie PendingModifiedValues). I could just check for those values separately though after your code. What do you think about that approach?

Redshift does allow additional modifies, it will just overwrite any previous pending modifies.

Reply:
thats why cluster is the default value for any values not in cluster_pending

also status fields are not something that can be changed i would think

Reply:
No I mean you would have to do `cluster.get(HsmStatus).get(HsmClientCertificateIdentifier)` for those values, since the key is not just in cluster. The status fields contain both the field and the status as separate keys, so cluster.ElasticIpStatus.ElasticIp can be modifed, but cluster.ElasticIpStatus.status can not be.

Reply:
I just pushed a change along the lines of what I was thinking